server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,185 @@
+            CacheStatsResponse::new
+        );
+        this.map = new ConcurrentHashMap<Key, CacheStatsResponse>(readMap);
+        this.totalStats = new CacheStatsResponse(in);
+        this.maxDimensionValues = in.readVInt();
+    }
+
+    public List<String> getDimensionNames() {
+        return dimensionNames;
+    }
+
+    public ConcurrentMap<Key, CacheStatsResponse> getMap() {
+        return map;
+    }
+
+    public CacheStatsResponse getTotalStats() {
+        return totalStats;
+    }
+
+    public void incrementHitsByDimensions(List<CacheStatsDimension> dimensions) {
+        internalIncrement(dimensions, (response, amount) -> response.hits.inc(amount), 1);
+    }
+
+    public void incrementMissesByDimensions(List<CacheStatsDimension> dimensions) {
+        internalIncrement(dimensions, (response, amount) -> response.misses.inc(amount), 1);
+    }
+
+    public void incrementEvictionsByDimensions(List<CacheStatsDimension> dimensions) {
+        internalIncrement(dimensions, (response, amount) -> response.evictions.inc(amount), 1);
+    }
+
+    public void incrementMemorySizeByDimensions(List<CacheStatsDimension> dimensions, long amountBytes) {
+        internalIncrement(dimensions, (response, amount) -> response.memorySize.inc(amount), amountBytes);
+    }
+
+    public void incrementEntriesByDimensions(List<CacheStatsDimension> dimensions) {
+        internalIncrement(dimensions, (response, amount) -> response.entries.inc(amount), 1);
+    }
+
+    public void decrementEntriesByDimensions(List<CacheStatsDimension> dimensions) {
+        internalIncrement(dimensions, (response, amount) -> response.entries.inc(amount), -1);
+    }
+
+    /**
+     * Reset number of entries and memory size when all keys leave the cache, but don't reset hit/miss/eviction numbers
+     */
+    public void reset() {
+        for (Key key : map.keySet()) {
+            CacheStatsResponse response = map.get(key);
+            response.memorySize.dec(response.getMemorySize());
+            response.entries.dec(response.getEntries());
+        }
+        totalStats.memorySize.dec(totalStats.getMemorySize());
+        totalStats.entries.dec(totalStats.getEntries());
+    }
+
+    public long count() {
+        // Include this here so caches don't have to create an entire CacheStats object to run count().
+        return totalStats.getEntries();
+    }
+
+    private void internalIncrement(List<CacheStatsDimension> dimensions, BiConsumer<CacheStatsResponse, Long> incrementer, long amount) {
+        CacheStatsResponse stats = internalGetStats(dimensions);
+        incrementer.accept(stats, amount);
+        incrementer.accept(totalStats, amount);
+    }
+
+    private CacheStatsResponse internalGetStats(List<CacheStatsDimension> dimensions) {
+        assert dimensions.size() == dimensionNames.size();
+        CacheStatsResponse response = map.get(new Key(dimensions));
```

Comment:
This looks a bit weird to me. Why are we treating a list of dimension as one key? Shouldn't we iterate over each dimension value and gets corresponding statsResponse object?
If I pass two dimension as part of ICacheKey
`"dimension1", dimension2"`

I am guessing we will track hits/misses separately for each dimension?

hits = {("dimension1", 4), ("dimension2", 5)}

Let me know if I am missing anything here.

Reply:
@peteralfonsi 
Okay.
I guess we can give that option to the user whether they want to treat each dimension separately or as one as a whole?
I am also looking for other opinions on this @msfroh